typhos/jira.py

```diff
@@ -0,0 +1,178 @@
+import getpass
+import logging
+import platform
+import socket
+import urllib
+import urllib.parse
+import urllib.request
+from typing import Dict
+
+from pydm.exception import raise_to_operator
+from qtpy import QtWidgets
+
+import typhos
+
+from . import utils
+
+SYSTEM_UNAME_DICT = dict(platform.uname()._asdict())
+logger = logging.getLogger(__name__)
+
+
+def _failsafe_call(func, *args, value_on_failure=None, **kwargs):
+    try:
+        return func(*args, **kwargs)
+    except Exception as ex:
+        if value_on_failure is None:
+            return f'FAILURE: {type(ex).__name__}: {ex}'
+        return value_on_failure
+
+
+class TyphosJiraIssueWidget(QtWidgets.QFrame):
+    _encoding = "utf-8"
+
+    def __init__(self, device=None, parent=None):
+        super().__init__(parent=parent)
+
+        layout = QtWidgets.QFormLayout()
+        layout.setFieldGrowthPolicy(layout.AllNonFixedFieldsGrow)
+        self.setLayout(layout)
+
+        self._submitted = False
+
+        self.name = QtWidgets.QLineEdit(getpass.getuser())
+        layout.addRow("Your &name", self.name)
+        # TODO jira suffix
+        self.email = QtWidgets.QLineEdit(
+            f"{getpass.getuser()}{utils.JIRA_EMAIL_SUFFIX}"
+        )
+        layout.addRow("Your &e-mail", self.email)
+        self.summary = QtWidgets.QLineEdit("")
+        layout.addRow("Issue &summary", self.summary)
+        self.details = QtWidgets.QPlainTextEdit("""\
+* What were you trying to do?
+
+* What did the device/interface/etc do?
+
+* What should have happened?
+
+* Please provide additional context here:
+        """.strip())
+
+        layout.addRow("Issue &details", self.details)
+        self.submit = QtWidgets.QPushButton("Submit")
+        layout.addRow(self.submit)
+        self.submit.clicked.connect(self._check_submission)
+        self.status = QtWidgets.QLabel()
+        layout.addRow(self.status)
+
+        self.device = device
+        if device is not None:
+            self.summary.setText(
+                f"Device {device.name} ({device.__class__.__name__})"
+            )
+            self.setWindowTitle(f"Typhos issue reporting ({device.name})")
+        else:
+            self.setWindowTitle("Typhos issue reporting")
+
+    @staticmethod
+    def get_environment():
+        """Get the default environment information."""
+        env = dict(
+            typhos=typhos.__version__,
+            user=getpass.getuser(),
+            hostname=_failsafe_call(socket.getfqdn),
```

Comment:
I think we also want some information on which Python env we're using, I think the most site-agnostic way would be something like:

`python_env=sys.base_prefix`

Which looks like:
```
In [3]: sys.base_prefix
Out[3]: '/reg/g/pcds/pyps/conda/py36/envs/pcds-4.1.6'
```

Reply:
Added in 70cd5e8

Reply:
Oh, we probably also want the `PYTHONPATH` environment variable? Or is that too much?

Reply:
Couldn't hurt... probably! I guess we'll see if there's a limit to the length the issue collector will accept.

Reply:
Adding that long list necessitated some formatting tweaks, which ended up making this better overall. Viewed from Jira:
<img width="463" alt="image" src="https://user-images.githubusercontent.com/5139267/126538086-2b965927-eb33-4669-b5ae-e48ca3177a1f.png">